Neuron package with cmake build system

iomaganaris opened the pull request:
Made the `NEURON` package a spack `CMakePackage` to be able to easily built the simulator with the `cmake` build system, while keeping compatibility with the `autotools` build system.
I was developing the recipe on my workstation and I tested it with `neurodamus-neocortex` and `neurodamus-python` and it worked fine. 
I also tested compiling mod files with `nrnivmodl` and `build_neurodamus.sh` and both worked fine.

TODO:

- [x] Test spack recipe in mac
- [x] Add proper neuron tag and version number
- [x] Enable cmake build system based on the version of neuron (make cmake default and conflict with versions [:7.8.0b]
- [x] Test the package in bb5 with various neurodamus versions
- [x] Test nrnivmodl, etc
- [x] Reorder and clean all the functions with `@when('+/-cmake')`
- [x] Fix issue with neuron-direct test

iomaganaris committed:
Cmake build system working and passing tests

iomaganaris committed:
[WIP] Neuron+cmake integration to Neurodamus

ferdonline committed:
Adding mvdtool 2.3.2, py-neurodamus, highfive/mock fix

ferdonline committed:
Fixes

iomaganaris committed:
Small fixes in neuron recipe and sim-model fixes for libnrnmech path

iomaganaris committed:
Added py-cython dependency in neuron

iomaganaris committed:
Made coreneuron variant False as default in neuron package

iomaganaris committed:
Added +debug variant in neuron

iomaganaris committed:
Added patch for reverting free_fast_imem to check the simstack

iomaganaris committed:
Reverted temporary patch for neuron. Now using as preferred pure pramodk/cmake_master_merge as version 8.0.0

iomaganaris committed:
Made cmake build false by default to check the autotools build system

iomaganaris committed:
Fixing filter_compilers

iomaganaris committed:
Reverted cmake as default to test changes in filter_compilers with cmake

iomaganaris committed:
Added dependency to py-test when python and tests are enabled

iomaganaris committed:
Clearing and fixing paths of neuron archdir

iomaganaris committed:
Edit neuron and sim-model/neurodamus-core packages with new variables for neuron_basedir
and nrnivmodl_outdir

iomaganaris committed:
Made cmake false by default to check the CI tests

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -210,6 +327,7 @@ def build_nmodl(self, spec, prefix):
         make()
```
I think this is dangerous w.r.t. build failures. Is there any way to separate `make` and `make install`?

iomaganaris replied:
You are right. I think it's possible to refactor a bit the `autotools` build and add all the `make()` calls at least in the `build` stage.

matz-e reviewed (commented):
Haven't looked too much in detail, but have one question already. Quite a bit of work here!

iomaganaris committed:
Fixed failing asciitoh5 package with neuron_archdir

iomaganaris committed:
Reorganized functions

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -27,7 +27,8 @@ class Neuron(Package):
 
     version('develop', branch='master')
-    version('7.8.0a',  commit='92a208b', preferred=True)
+    version('8.0.0',   branch='master', preferred=True)
```
If 8 is the latest version, the `preferred` is redundant and should be dropped.

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -37,92 +38,120 @@ class Neuron(Package):
-    variant('shared',        default=True,  description='Build shared libraries')
-    variant('pysetup',       default=True,  description="Build Python module with setup.py")
-    variant('rx3d',          default=False, description="Enable cython translated 3-d rxd. Depends on pysetup")
+    variant('cmake',                   default=False, description="Build NEURON using cmake")
```
I'd prefer to drop the extra whitespace to align the arguments, will make future diffs smaller.

iomaganaris replied:
I removed a large portion of them but kept some to look nicer, similar to what is done in other projects. Let me know what you think :+1:

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -37,92 +38,120 @@ class Neuron(Package):
+        def cmake_enable_option(name, spec_requiremement):
+            value = 'TRUE' if spec_requiremement in self.spec else 'FALSE'
+            return '-DNRN_ENABLE_' + name + ':BOOL=' + value
+        args = [
```
Nice! Could maybe be improved with
`args = [cmake_enable_option(setting, variant) for (setting, variant) in [('SHARED', '+shared'), (…)…]]`. Maybe…

iomaganaris replied:
I refactored this a bit to avoid writing `SHARED` for `+shared`, etc. Check it out if you want

matz-e reviewed (commented):
Looks good from my side so far! Minor optional suggestion included.

iomaganaris committed:
Addressed some comments from @matze. Thank you!

- Removed whitespaces from variants
- Fixed preferred version

iomaganaris committed:
Fixes in versions

- Reverted back explicitly setting the preferred version otherwise 2018-10 is used
- Set version 7.8.1 to point to the fix for load balance (with the cmake commit) [subjected to change]

tristan0x commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -25,14 +25,13 @@ class Neuron(Package):
 
     version('develop', branch='master')
-    version('7.8.0b',  commit='92a208b', preferred=True)
+    version('7.8.1',   commit='79a4d2a', preferred=True)
```
Latest is always preferred.
```suggestion
    version('7.8.1',   commit='79a4d2a')
```

iomaganaris replied:
Good point but in this case `preferred` should be defined explicitly otherwise version `2018-10` is selected

tristan0x commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -42,92 +41,120 @@ class Neuron(Package):
+    variant('binary',     default=False, description="Create special as a binary instead of shell script")
+    variant('coreneuron', default=False, description="Patch hh.mod for CoreNEURON compatibility")
+    variant('cross-compile',  default=False, description='Build for cross-compile environment')
+    variant('debug',          default=False, description='Build debug with O0')
```
```suggestion
    variant('debug',          default=False, description='Build with debugging information and without optimization')
```

or 

```suggestion
    variant('debug',          default=False, description='Build with flags -g -O0')
```

iomaganaris replied:
I think I prefer the second version, since it is more explicit. Thanks for the suggestions :+1:

tristan0x commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -42,92 +41,120 @@ class Neuron(Package):
+    variant('shared',     default=True,  description='Build shared libraries')
+    variant('tests',      default=False, description='Enable unit tests')
+    variant('threads',    default=True,  description='Allow use of Pthreads')
+    variant('debug',      default=False, description='Build debug with O0')
```
The _debug_ variant is duplicated.

iomaganaris replied:
Oh god.. Thanks for pointing that out :+1:

tristan0x reviewed (dismissed):
That is a very nice patch! Well done @iomaganaris 
I only have minor remarks.

iomaganaris committed:
Fixed CMAKE_BUILD_TYPE when +cmake+debug

iomaganaris committed:
Removed duplicate debug variant and rephrased debug decription according to Tristan's suggestion

ohm314 commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -42,92 +41,120 @@ class Neuron(Package):
+    depends_on('py-cython',   when='+python', type=('build', 'link', 'run'))
     depends_on('tau',         when='+profile')
 
+    conflicts('+cmake',   when='@0:7.8.0b')
```
```suggestion
    conflicts('+cmake',   when='@:7.8.0b,2018-10')
```

ohm314 commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -42,92 +41,120 @@ class Neuron(Package):
+    # == Autotools build system related functions ==
+    # ==============================================
+    @when('~cmake')
+    @run_before('build')
```
Nice, I like it!

ohm314 reviewed (dismissed):
Looks great, thank you! Please see just one minor remark.

iomaganaris committed:
Small changes to filter_compilers to filter files according to the build system

iomaganaris committed:
Added version 2018-10 of Neuron to the conflicting versions with cmake

iomaganaris committed:
Refactored cmake_args and cmake_enable_option a bit

iomaganaris committed:
Removed filter_compier_wrappers for files that don't exist from cmake_args and set_autoconf_options

pramodk commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -39,92 +40,112 @@ class Neuron(Package):
+    variant('multisend',  default=True,  description="Enable multi-send spike exchange")
+    variant('profile',    default=False, description="Enable Tau profiling")
+    variant('pysetup',    default=True,  description="Build Python module with setup.py")
+    variant('python',     default=True,  description='Enable python')
```
I would just combine above two options as python.

iomaganaris replied:
`pysetup` is only for the `autotools` build, right? With `cmake` the `python` package is built by default. Also, regarding the `dynamic python version support`, I guess that there is no usage in spack?

pramodk commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -39,92 +40,112 @@ class Neuron(Package):
+    variant('coreneuron', default=False, description="Patch hh.mod for CoreNEURON compatibility")
+    variant('cross-compile',  default=False, description='Build for cross-compile environment')
+    variant('debug',          default=False, description='Build with flags -g -O0')
+    variant('discrete-event-observer', default=True,  description='Enable Observer to be a subclass of DiscreteEvent')
```
I would remove this. I don’t see anyone using this variant practically.

pramodk commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -39,92 +40,112 @@ class Neuron(Package):
+    variant('rx3d',       default=True,  description="Enable cython translated 3-d rxd. Depends on pysetup")
+    variant('shared',     default=True,  description='Build shared libraries')
+    variant('tests',      default=False, description='Enable unit tests')
+    variant('threads',    default=True,  description='Allow use of Pthreads')
```
Same...we can remove this as practically we won’t be using this.

pramodk commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -39,92 +40,112 @@ class Neuron(Package):
+    depends_on('automake',  type='build', when='~cmake')
+    depends_on('bison',     type='build')
+    depends_on('flex',      type='build')
+    depends_on('libtool',   type='build')
```
Libtool if for ~cmake.

pramodk reviewed (changes requested):
Very nice! I added Minor comments/questions m.

iomaganaris committed:
Addressed Pramod's comments

- Removed discrete-event-observer variant
- Removed threads variant
- Merged pysetup and python variants

iomaganaris committed:
Fixed issue with ImportError: No module named site in mac os

iomaganaris committed:
Made '+coreneuron' default and set MPICXX_CXX

iomaganaris committed:
CMake build fixes

- Added clean=False in version 7.8.1 because it is based in a commit
  and spack deletes .git folder when it creates the archive for it
  and cmake build fails if it uses this archive again
- made cmake variant True by default

iomaganaris committed:
Added ~cmake in older versions of neuron that exist in deployment

iomaganaris committed:
Added coreneuron-compatible variant
- When enabled install Neuron with patched hh.mod file for CoreNeuron
  compatibility without installing CoreNeuron as submodule of Neuron
- If CoreNeuron is installed as submodule of Neuron and as spack package
  there are issues when selecting which CoreNeuron to use in neurodamus
  installation

iomaganaris committed:
Revert "Added coreneuron-compatible variant"

This reverts commit d8c42c4de3f9d0a13f201f567bfcf161058d80e9.

iomaganaris committed:
Install neuron+coreneuron by default and then remove CoreNeuron executables

- This is needed to keep compatibility of the output of Neuron with CoreNeuron
  (basically edit properly the hh.mod file)

iomaganaris committed:
Attempt to fix issue with numpy when installing Neuron in CoreNeuron jenkins plan

iomaganaris committed:
Bumped testing version to 7.8.2 to avoid reusing old archive in the CI

- Also added clean=False to version 7.8.0b which is based in a commit

iomaganaris commented:
CoreNeuron CI https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/oss.coreneuron.pipeline/detail/oss.coreneuron.pipeline/766/pipeline/ and SimulationStack CI https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/1163/pipeline/ passed, so this PR is ready for review.
The only thing missing/that needs discussion is how to pick the preferred `Neuron` version that uses the `cmake` build system.

iomaganaris committed:
Set git variable in CoreNeuron spack recipe

iomaganaris committed:
Refactored deletion of installed coreneuron binaries

iomaganaris committed:
Merge branch 'develop' into pr/neuron-cmake

iomaganaris committed:
Fixed installation after merge with develop in mac

- Fixed issue with clang compiler in filter_file
- Fixed issue with libnrnmech.dylib not found

iomaganaris committed:
Added mod-compatibility flag for patching mod files for coreneuron

iomaganaris committed:
Fix dylib

iomaganaris committed:
Updated neuron versions

iomaganaris committed:
flake8 fixes

iomaganaris committed:
Change find to find_libraries when finding libnrnmech

iomaganaris committed:
Fix issue with clang/clang++ compiler and add conflict for neuron+binary+cmake

iomaganaris committed:
Updated Neuron version and attempt to fix Travis CI issues

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -435,12 +435,12 @@ def filter_compilers(self):
-        filter_file('CXX = {}'.format(env["CXX"]),
-                    'CXX = {}'.format(cxx_compiler),
+        filter_file('CXX = %s' % env["CXX"],
+                    'CXX = %s' % cxx_compiler,
```
Why????? `%` has to die!¡
```suggestion
                    'CXX = {0}'.format(cxx_compiler),
```

iomaganaris replied:
I thought that there was an issue about that with travis.. I will fix it and change it back to `format`. Thanks 🙂

matz-e replied:
Thank you! `{}` was introduced in 2.7, but `{0}` et al are ok with Python 2.6, which is what the CI should test…

iomaganaris replied:
I see.. Thank you for the explanation! 🙂

iomaganaris committed:
Change % to format for string formatting

iomaganaris committed:
Fix for finding libnrnmech.so.0.0.0 with neuron~cmake

pramodk commented:
I just realised one thing : CMake build system on NEURON needs support for MPI dynamic i.e. we shouldn't link MPI libraries to libnrniv.

pramodk commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -52,138 +53,136 @@ class Neuron(Package):
-        "+mpi+multisend": ["--with-multisend"],
-        "~rx3d": ["--disable-rx3d"],
-        "~mpi": ["--without-paranrn"],
-        "+mpi": ["--with-paranrn=dynamic"],
```
Note that the equivalent option for CMake is `-DNRN_ENABLE_MPI_DYNAMIC=ON`.

iomaganaris committed:
Merge branch 'develop' into pr/neuron-cmake

iomaganaris committed:
Update version 7.8.0c to latest master of neuron and fix patches

iomaganaris committed:
Added latest reportinglib tag

iomaganaris committed:
Enable dynamic MPI with cmake

iomaganaris commented:
please retest

matz-e commented:
> please retest

You need to be nicer and not hurt the CIs feelings.

Retest this please

matz-e commented:
The shorter phrase I added to our Jenkins at some point. Should have paid more attention and made it the same as the standard phrase. Also Jenkins default…

iomaganaris commented:
> > please retest
> 
> You need to be nicer and not hurt the CIs feelings.
> 
> Retest this please

As you mentioned I thought that the standard phrase also worked here 😕 
I ❤️ Jenkins

iomaganaris committed:
Fixed asciitoh5 package

iomaganaris commented:
Retest this please

iomaganaris commented:
https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/1727/pipeline
Simulation Stack passed

iomaganaris committed:
Updated to latest commit

iomaganaris committed:
Fix issue with py-ipykernel test

iomaganaris commented:
> Didn’t understand why pay-backcall change is required here.

https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.spack-deployment.pull-requests/detail/hpc.spack-deployment.pull-requests/569/tests
The `py-ipykernel` tests fail with:
```
ModuleNotFoundError: No module named 'backcall'
```

matz-e commented:
> > Didn’t understand why pay-backcall change is required here.
> 
> https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.spack-deployment.pull-requests/detail/hpc.spack-deployment.pull-requests/569/tests
> The `py-ipykernel` tests fail with:
> 
> ```
> ModuleNotFoundError: No module named 'backcall'
> ```

It's a transitive, not a direct dependency, though: https://github.com/ipython/ipykernel/blob/master/setup.py#L91

iomaganaris commented:
> > > Didn’t understand why pay-backcall change is required here.
> > 
> > 
> > https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.spack-deployment.pull-requests/detail/hpc.spack-deployment.pull-requests/569/tests
> > The `py-ipykernel` tests fail with:
> > ```
> > ModuleNotFoundError: No module named 'backcall'
> > ```
> 
> It's a transitive, not a direct dependency, though: https://github.com/ipython/ipykernel/blob/master/setup.py#L91

True. I see that your PR for `py-ipython` passes the tests, so I will rebase once this is merged on `develop`.
Thank you!

iomaganaris committed:
Revert "Fix issue with py-ipykernel test"

This reverts commit 18d23d3e047a2ef963f7a2268b346d1052447a11.

iomaganaris committed:
Merge branch 'develop' into pr/neuron-cmake

iomaganaris committed:
Merge branch 'develop' into pr/neuron-cmake

pramodk commented:
please retest

pramodk commented:
Retest this please

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -24,15 +24,16 @@ class Neuron(Package):
 
     # Patch which reverts 81a7a39 for numerical compat
-    patch("revert_Import3d_numerical_format.master.patch", when="@develop")
+    patch("revert_Import3d_numerical_format.master.patch", when="@7.8.0c:develop")
```
```suggestion
    patch("revert_Import3d_numerical_format.master.patch", when="@7.8.0c:")
```
Since the version range is inclusive, might as well just make this open-ended.

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -24,15 +24,16 @@ class Neuron(Package):
-    version("7.6.8", tag="7.6.8")
-    version("7.6.6", tag="7.6.6")
+    version("7.8.0c",  commit="81f9e60", preferred=True)
+    version("7.8.0b",  commit="92a208b")
```
ugh… these are beta versions?

pramodk replied:
Versioning scheme is inconsistent as we don't have freedom to fix stuff upstream. Fernando is fixing that. For now we will use `7.8.0c`.

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -52,138 +53,136 @@ class Neuron(Package):
-        default="1",
-        description="Build number for re-builds",
-    )
+    variant("deployment_build", default="1",  description="Build number for re-builds")
```
We should drop this for a new version.

pramodk replied:
done

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -52,138 +53,136 @@ class Neuron(Package):
+    variant("deployment_build", default="1",  description="Build number for re-builds")
+
+    variant("codechecks", default=False,
+            description="Perform additional code checks like " +
```
```suggestion
            description="Perform additional code checks like "
```

pramodk replied:
done

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -330,61 +402,90 @@ def filter_compilers(self):
-        )
+        libtool_makefile = join_path(self.prefix,
+                                     self.basedir,
+                                     "../share/nrn/libtool")
```
The old way seems more Pythonic to me here… what would Black do?

pramodk replied:
Done

matz-e reviewed (commented):
Some minor comments.

I also see a little C++ style creep here with a lot of things aligned to the right rather than just indented after linebreaks. IMO that makes Python code harder to read.

iomaganaris committed:
Added cmake option for building binary special

iomaganaris committed:
Some more additions

iomaganaris committed:
Fixed binary option

iomaganaris committed:
Fix libnrnmech lib name with neuron+cmake+binary

iomaganaris committed:
Change binary special cmake option

iomaganaris committed:
Merge branch 'develop' into pr/neuron-cmake

iomaganaris committed:
Update neuron to latest master

pramodk committed:
Address review comments

pramodk reviewed (commented):
Almost ready!

pramodk committed:
Merge branch 'develop' into pr/neuron-cmake

pramodk committed:
flake8 fixes

iomaganaris committed:
Fixed filtering of CC/CXX values in nrnmech_makefile

iomaganaris committed:
Set CC/CXX env variables to mpicc/cxx in dependent environments of nrn

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -471,14 +471,22 @@ def setup_run_environment(self, env):
             env.set("MPICC_CC", self.compiler.cc)
             env.set("MPICXX_CXX", self.compiler.cxx)
+            env.set("CC", self.spec["mpi"].mpicc)
+            env.set("CXX", self.spec["mpi"].mpicxx)
```
This looks like those two are swapped? What's the difference?

iomaganaris replied:
`MPICC_CC` sets the underlying compiler for `mpicc` to the gcc used to install the package instead of the one coming from the system when `neuron` module is loaded and by setting `CC` to `mpicc` we compile with the mpi libraries. Setting `CC` is needed because `nrnivmodl` which compiles the `mod files` looks for the `CC/CXX` variables set in the environment for the compilation of the `mod files`

matz-e replied:
Thanks!

pramodk committed:
Change 7.8.0c to latest fix with binary special

pramodk committed:
Merge branch 'develop' into pr/neuron-cmake

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -30,7 +30,7 @@ class Neuron(CMakePackage):
     patch("apply_79a4d2af_load_balance_fix.patch", when="@7.8.0b")
 
     version("develop", branch="master")
-    version("7.8.0c",  commit="defabcd", preferred=True)
```
Did you clean out the PR directory?

matz-e replied:
Also, that was a nice commit hash :(

pramodk commented:
Retest this please

pramodk committed:
Changes to env variables and filter file
 - for darwin, it shouldn't be necessary to unset PYTHONPATH
 - CC/CXX should alway point to mpicc/mpicxx rather than
    allowing user to override it. This is required because
    imagine if user is doing `module load neuron gcc` vs.
    `module load gcc neuron` will result into different
    behaviour! We want to avoid this in porudction.
  - cython is required for rx3d only as build dependency

pramodk commented:
@iomaganaris : I explained in [this commit](https://github.com/BlueBrain/spack/pull/644/commits/6d5e58afe86431334332ecb83c8da510943bb241) my changes to neuron recipe. We can fix the issues if we see failures.

One thing that I am noticing is that with CMake neuron when I directly use binary:

```
/gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/ringtest
ls: cannot access mod//*.inc: No such file or directory
mod//halfgap.mod
halfgap.mod
 -> Compiling mod_func.c
 -> NMODL halfgap.mod
Translating halfgap.mod into halfgap.c
Thread Safe
 -> Compiling x86_64/halfgap.c
 => LINKING library x86_64/libnrnmech.so Mod files: halfgap.mod
 => LINKING executable x86_64/special LDFLAGS are:    /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/readline-8.0-uuhvgs2hgl/lib/libreadline.so -Wl,-rpath,/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/readline-8.0-uuhvgs2hgl/lib /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib/libpython3.7m.so -Wl,-rpath,/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/25-05-2020/install/linux-rhel7-x86_64/gcc-8.3.0/neuron-7.8.0c-chquzd/lib/libnrniv.so: undefined reference to `VTT for std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/25-05-2020/install/linux-rhel7-x86_64/gcc-8.3.0/neuron-7.8.0c-chquzd/lib/libnrniv.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::rfind(char const*, unsigned long, unsigned long) const@GLIBCXX_3.4.21'
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/25-05-2020/install/linux-rhel7-x86_64/gcc-8.3.0/neuron-7.8.0c-chquzd/lib/libnrniv.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)@GLIBCXX_3.4.21'
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/25-05-2020/install/linux-rhel7-x86_64/gcc-8.3.0/neuron-7.8.0c-chquzd/lib/libnrniv.so: undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)@GLIBCXX_3.4.20'
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/25-05-2020/install/linux-rhel7-x86_64/gcc-8.3.0/neuron-7.8.0c-chquzd/lib/libnrniv.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find(char, unsigned long) const@GLIBCXX_3.4.21'
```

This is expected because if compiler is not loaded then nrnivmodl will be using system compiler which has different ABI than gcc/8.3.0. But, this was working fine with autoconf because in CMake we have enabled C++11 (even if neuron is currently not using it).

~We can check if somehow we can make it compatible.~ I think as long as people load neuron module (where we set default MPICC_CC and MPICXX_CXX), it should work fine.

pramodk commented:
Retest this please

iomaganaris committed:
Set PYTHONPATH for run env

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -446,12 +446,24 @@ def filter_compilers(self):
+        for pydir in (    
+                self.spec.prefix.lib64.python,    
+                self.spec.prefix.lib.python,    
+        ):    
```
Embrace the sadface!

matz-e replied:
Probably could be outdented by four spaces… fine though if flake8 passes

iomaganaris replied:
You're right, missed that and a couple of trailing whitespaces coming from copy paste. It should be fine now 🙏

pramodk replied:
I think Matthias will be happy now  with https://github.com/BlueBrain/spack/pull/644/commits/0e2f82ffc09693013b479ecb3511795dabd30131

iomaganaris committed:
Make flake8 happy

pramodk committed:
Use self.spec.prefix.lib.python for PYTHONPATH

pramodk committed:
Try removing setup_dependent_build_environment

pramodk committed:
use setup_dependent_run_environment as well for PYTHONPATH

pramodk reviewed (approved):
All previous review comments are addressed and these changes are tested.
So I will go ahead with the merge.
